### ISSUE TITLE
OpenSSL 1.1.1h is no longer reporting expired root CAs when using "verify"

### ISSUE THREAD

**Issue by ronf** [openssl/openssl#13471]
I've run into an issue with OpenSSL 1.1.1h when trying to verify a certificate chain that ends at an expired root certificate. The expiration check worked fine in OpenSSL 1.1.1g, properly reporting an expired root CA. However, with OpenSSL 1.1.1h, it reports that the verification succeeded even though the root certificate was expired.

Here's what I used to reproduce this issue:

Example root certificate saved as root.pem:

```
-----BEGIN CERTIFICATE-----
MIICzzCCAbegAwIBAgIUOOQ+yEDsApy5V50CVjbyRg+Jb54wDQYJKoZIhvcNAQEL
BQAwDzENMAsGA1UEAwwEcm9vdDAeFw0yMDExMTkyMzA1MjJaFw0yMDExMjAyMzA1
MjJaMA8xDTALBgNVBAMMBHJvb3QwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEK
AoIBAQDwtPzP8MvxlC7uH1acWI1gRt5IYJXYejLJ5aGdSTp7WpeDu7ZpSusMbwcs
L7Vxydo5KCnEn+HhWxiBnuwBp43RgNVrTIFHFIHP+nsE+xnKDp7OG/T1zIuyLH5r
iCulyJ6ss3V0T3bHkFzWP7eN3gFsznOeYy09pZ4oHqbUv8CGuBE5J8PxdFiGj0ya
0LGbT7/Rj9HoOzyPCXtnuAAxWsusD6fmDJOhK0bZkimYJkgR17IqrMYbGu1Lao+L
9f8ZARJsgOEzzoWvze3/V2RVnCjnL+ecwdiuvCrrazcBs1oUyGjcy4KjfLxF0jrS
AVuWJZW/k8FVVcLSX0AOPHsCNveXAgMBAAGjIzAhMA8GA1UdEwEB/wQFMAMBAf8w
DgYDVR0PAQH/BAQDAgEGMA0GCSqGSIb3DQEBCwUAA4IBAQAgMFXQ3JbCDWvVrNyf
LHCVETn04lVjKKdXR1aTHqglw0ecIfmUgkca6weby78iCvAE1BexkafZsEaeehob
a0myk40NyTU/iPsgVUm/KvGNzMUeFW4kHzFzu50ilEovE5W/jabQZzQCXG9O8A6r
pztots2WX53BeUqX2iMQvT2/LFPMUYKOwPv3toLdHVAMwuMSsmI8DRgQ8klMd4Sa
bsxJL7oXDPdE7LahRJgEvNqye/CdumiGthxhFod/+BdKKK9zruC43TPlxlYgG+EZ
vidiApo4gXhvTecZHS5bWMnHQ/9QbcQnIDd7Nqm2hkWCSgY1Yvph1ex2oNNotXK0
odXK
-----END CERTIFICATE-----
```

Example server certificate saved as server.pem:

```
-----BEGIN CERTIFICATE-----
MIICzjCCAbagAwIBAgIUHiCc6FlLb4cAmSk96GPLGNwlfe4wDQYJKoZIhvcNAQEL
BQAwDzENMAsGA1UEAwwEcm9vdDAeFw0yMDExMTkyMzI5MzNaFw0yMTExMjEyMzI5
MzNaMBExDzANBgNVBAMMBnNlcnZlcjCCASIwDQYJKoZIhvcNAQEBBQADggEPADCC
AQoCggEBALruu/SthRMZKpk8oCLA29NBxV361OQETZHuoliuru4IgEEL33b4XzFw
U0Ay06tPvPenYy4sznMzbwAz9DsjPGsBo1BNXqyNhd23k8cl4A+s/pwQNtK/WZoM
VR6s3klaWFMvhZY/oStWK/KVLuJZDPPuJNXBgE6gU24BRmb1sckQNILMqVb8qmJY
70O6iYbZHKs3c+t2P93IKbT9ownS3017pUBKiavpAYT3i4SHJF2A7gjmBoWBs9zp
aiJjhgqB5xyzuoeMYv+DovCOSz6katr8NB9MnLMI5pGCqt/F2cwFHmF4cngeD74w
Igp2VdSGc1B18V0bFS3kRkko3vFtLssCAwEAAaMgMB4wDAYDVR0TAQH/BAIwADAO
BgNVHQ8BAf8EBAMCA6gwDQYJKoZIhvcNAQELBQADggEBAC1psVLCrti+LmipwZ1J
v/ZE4B4GzBwnifRRkKT+CIJ+XYmmQlljVdnpjcs1bIQ11K7yQjMF5ZVOQBFHYo5h
zvRZUsMbs0NsxRYBhE2m9xxTMe4ybDZhhSlaJnt6HHWUYxtVr6NAPvgUlYkoIBhh
UlIR7BuyQwOJVburbNMc0UdBJGI71RfP/fsk50Cf12YcBKqfayeeLDj9dS71SmQ9
9do6YiOfB8wGFX+p+iKuu7I0iPJNmGtRacwVKPlDzldtrxrDapNEDFx0cFsvwkSC
WPQvuLhveCx7lZNqWTpVaHJShOEsJR65xdn8Jk2cSzpfLwO/felsfTNv1OraDvY0
MrI=
-----END CERTIFICATE-----
```

Summary of decoded root CA, intentionally kept as bare-bones as possible:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            38:e4:3e:c8:40:ec:02:9c:b9:57:9d:02:56:36:f2:46:0f:89:6f:9e
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = root
        Validity
            Not Before: Nov 19 23:05:22 2020 GMT
            Not After : Nov 20 23:05:22 2020 GMT      <-- Expired
        Subject: CN = root
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    <snip>
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
    Signature Algorithm: sha256WithRSAEncryption
         <snip>
```

Summary of decoded server cert:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            1e:20:9c:e8:59:4b:6f:87:00:99:29:3d:e8:63:cb:18:dc:25:7d:ee
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = root
        Validity
            Not Before: Nov 19 23:29:33 2020 GMT
            Not After : Nov 21 23:29:33 2021 GMT
        Subject: CN = server
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    <snip>
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment, Key Agreement
    Signature Algorithm: sha256WithRSAEncryption
         <snip>
```

Command to verify the certificate:

```
openssl verify --CAfile root.pem server.pem
```

Expected output, which was seen on OpenSSL 1.1.1g:

```
CN = root
error 10 at 1 depth lookup: certificate has expired
error server.pem: verification failed
```

Unexpected output seen on OpenSSL 1.1.1h:

```
server.pem: OK
```

I haven't completely tracked down the source of the problem, but it seems related to changes in the check_issued() function in x509/x509_vfy.c. In 1.1.1g this returns 1, but in 1.1.1h it returns 0, causing the internal_verify() function to stop after checking the server cert and not check the root cert.

**Comment by kaduk** [comment]
It seems we've had multiple issues relating to root certificate expiry, recently (#13214 #13427 and maybe more).  Note that it's not always clear that checking the expiration of the root CA is useful, though that discussion is probably a separate one from this question of a behavior change.
Looking at the commit history, #10587 looks to be touching things vaguely in this space; @DDvO does this issue description prompt any thoughts from you?

**Comment by DDvO** [comment]
Indeed the backport #10587, which is commit e2590c3a162eb118c36b09c2168164283aa099b4, caused this change of behavior, which was not intended.
I'll investigate further on this ASAP.

**Comment by kaduk** [comment]
Thank you!

**Comment by DDvO** [comment]
A fix is coming up in #13585.

**Comment by DDvO** [comment]
BTW, @ronf, did you notice that your example (root and) server cert is not RFC 5280 conformant because it does not include an AKID?

OpenSSL 3.0 will reveal this if the `-x509_strict` option is added for the verification:
```
apps/openssl verify -no_check_time -x509_strict -CAfile root.pem server.pem
```
yields
```
CN = server
error 85 at 0 depth lookup: Missing Authority Key Identifier
error server.pem: verification failed
```

**Comment by DDvO** [comment]
May I ask how you produced those example certs?
Unfortunately, the OpenSSL apps by default do not include subject key identifiers (SKID) nor issuer/authority key identifiers (AKID).

**Comment by ronf** [comment]
Thanks so much for looking into this, David!

Since this problem originally showed up as a unit test failure in my AsyncSSH package, I used it to generate certificates. This is a Python package, and it relies on the Python "cryptography" library, specifically cryptography.x509.CertificateBuilder. You can find the code which calls that at https://github.com/ronf/asyncssh/blob/master/asyncssh/crypto/x509.py#L303.

Regarding the AKID, I must admit I didn't realize that RFC 5280 made inclusion of AKIDs mandatory. Thanks for pointing that out. I'll look into adding that to my certificate generation function. Support for AKIDs and SKIDs is already supported by the cryptography library, so it should be quite straightforward.

I'm guessing that the AKID requirement would only apply to the child certificates, though, and that a self-signed root CA would only be expected to contain a SKID which matched the AKID in any certificates it generated. Does that sound right to you?

**Comment by DDvO** [comment]
Thanks @ronf for sharing details how those test certs were generated.

Yes, your thoughts on when an AKID and SKID are required are absolutely right. 
I've meanwhile provided pointers to the respecitve RFC 5280 subsections in #13603.

**Comment by ronf** [comment]
Thanks for the confirmation, @DDvO. I added this support to the AsyncSSH "develop" branch last night in ronf/asyncssh@18d9ad723bca8750a4a912ccf84e8c6094bbff07.

**Comment by DDvO** [comment]
Closing this since the fix meanwhile has been merged: #13585
and also a fix for the other issue that came up here is being handled: #13658.

**Comment by ronf** [comment]
Thanks very much! I have confirmed that OpenSSL 1.1.1i works as expected in my testing.
